### ISSUE TITLE
Excessive disk usage with ashift=16

### ISSUE THREAD

**Issue by cvoltz** [zfsonlinux/zfs#8163]
### System information

Type | Version/Name
 --- | --- 
Distribution Name	|  CentOS
Distribution Version	|  7.5
Linux Kernel	|  3.10.0-862.3.3
Architecture	|  x86_64
ZFS Version	|  0.7.9-1
SPL Version	|  0.7.9-1

### Describe the problem you're observing

With `ashift=16`, the available space reported by `zfs list` is 32% of the available space reported by `zpool list` in a 11 disk RAIDZ2 pool. Each drive is 4 TiB and has a physical block size of 4 KiB (so `ashift=12` is the appropriate setting).

I created a zpool using the same 11 disks and `recordsize=1M` and varied the value of `ashift` from 9 to 16 (the maximum allowed) to see how capacity reduced as a function of `ashift`. These are the results I got:

ashift | logical block size | used | available | refer | % of raw
-- | -- | -- | -- | -- | --
9 | 512 B | 162 KiB | 31.5 TiB | 48.8 KiB | 78
10 | 1 KiB | 237 KiB | 31.2 TiB | 62.8 KiB | 78
11 | 2 KiB | 407 KiB | 30.6 TiB | 123 KiB | 76
12 | 4 KiB | 740 KiB | 29.5 TiB | 219 KiB | 73
13 | 8 KiB | 1.39 MiB | 29.5 TiB | 439 KiB | 73
14 | 16 KiB | 2.44 MiB | 25.8 TiB | 767 KiB | 71
15 | 32 KiB | 4.87 MiB | 25.8 TiB | 1.50 MiB | 71
16 | 64 KiB | 4.86 MiB | 12.9 TiB | 1.49 MiB | 32

![image](https://user-images.githubusercontent.com/549284/49253319-fea6ab80-f3eb-11e8-858a-df2509e56ae9.png)

The values of `ashift` other than 16 are what I would expect based on capacity being lost to slack space but I can't rationalize the resulting capacity for `ashift=16`. It looks like a bug to me.

### Describe how to reproduce the problem

Run this script (changing the drives to match ones on your machine):
```bash
#!/bin/bash

rm -f test.log &> /dev/null

for ashift in `seq 9 16`; do
  echo -e "ashift:\t$ashift"
  echo -e "ashift:\t$ashift" &>> test.log

  zpool create ost02 \
    -o ashift=$ashift \
    -O canmount=off \
    -O recordsize=1M \
    -f \
    raidz2 \
      /dev/mapper/d6020_sep50014380412150BC_bay006-0 \
      /dev/mapper/d6020_sep500143804121153C_bay006-0 \
      /dev/mapper/d6020_sep500143804121453C_bay006-0 \
      /dev/mapper/d6020_sep5001438041210A3C_bay008-0 \
      /dev/mapper/d6020_sep50014380412150BC_bay007-0 \
      /dev/mapper/d6020_sep500143804121153C_bay007-0 \
      /dev/mapper/d6020_sep500143804121453C_bay007-0 \
      /dev/mapper/d6020_sep5001438041210A3C_bay009-0 \
      /dev/mapper/d6020_sep50014380412150BC_bay008-0 \
      /dev/mapper/d6020_sep500143804121153C_bay008-0 \
      /dev/mapper/d6020_sep500143804121453C_bay008-0 \
    &>> test.log

  echo -ne "zpool:\t"
  zpool list -H -o size,alloc,free ost02

  echo -ne "zfs:\t"
  zfs list -H -o used,avail,usedds ost02

  zpool destroy ost02 &>> test.log
  echo
done
```

The output looks like this:
```
ashift: 9
zpool:  40T     225K    40.0T
zfs:    151K    31.5T   48.8K

ashift: 10
zpool:  40T     354K    40.0T
zfs:    229K    31.2T   62.8K

ashift: 11
zpool:  40T     654K    40.0T
zfs:    407K    30.6T   123K

ashift: 12
zpool:  40T     1.22M   40.0T
zfs:    740K    29.5T   219K

ashift: 13
zpool:  40T     2.37M   40.0T
zfs:    1.39M   29.5T   439K

ashift: 14
zpool:  40T     4.59M   40.0T
zfs:    2.34M   25.8T   767K

ashift: 15
zpool:  40T     9.19M   40.0T
zfs:    4.68M   25.8T   1.50M

ashift: 16
zpool:  40T     18.9M   40.0T
zfs:    4.86M   12.9T   1.49M
```

**Comment by richardelling** [comment]
I believe this is as expected. The estimate for remaining space is based on an expected
recordsize=128k. So the ratio of recordsize:physical blocksize (via ashift) is 2:1. The rest
falls out of the analysis at:
https://www.delphix.com/blog/delphix-engineering/zfs-raidz-stripe-width-or-how-i-learned-stop-worrying-and-love-raidz
and 
https://docs.google.com/spreadsheets/d/1tf4qx1aMJp8Lo_R6gpT689wTjHv6CGVElrPqTA0w_ZY/edit?pli=1#gid=930519344

NB, ashift doesn't really need to be larger than the actual physical blocksize and you'll discover that little testing is done at ashift > 13.

**Comment by ahrens** [comment]
I agree with @richardelling.  This is behaving as designed.  Since (assuming the default recordsize=128K and large files), you are using blocks that are each 2 * ashift (2 "sectors" per ZFS's understanding), RAIDZ2 uses 2 parity + 2 data + 2 pad sectors = 6 sectors.  So we expect usable storage to be 1/3 the raw storage, which is exactly what you see.

In practice, if you use larger recordsize, you'll get more usable storage in practice (e.g. recordsize=1M -> 16 sectors per block -> usable space is 75% of raw).

**Comment by cvoltz** [comment]
I understand that `ashift` doesn't need to be larger then the actual physical block size and that little testing is done with values larger than 13 and I was familiar with the article and the spreadsheet but I wasn't sure what size `zfs list` was assuming. Does it always 128 KiB or does it really use the `recordsize`?

When I do the calculations for a write of 128 KiB of data I get:

```
# drives: 11 (9d + 2p)
# data blocks: 2
# parity blocks: 2
# padding blocks: 2
# blocks in stripe: 6
overhead (parity + padding): 4/6 ~ 67%
```

The number of drives and RAIDZ level are given parameters. With `ashift=16`, blocks are 64 KiB in size. When writing 128 KiB of data, we would be writing 2 data blocks (128 KiB/64 KiB). Since it is RAIDZ2 and the number of data blocks in the stripe is less than the number of data drives, we end up with 2 parity blocks. Since the data + parity is 4 blocks and that isn't a multiple of (# parity drives + 1 = 3), we add 2 padding blocks so the number of blocks in the stripe is 6. Since we are writing 6 blocks to write 2 blocks of data, the overhead is 4/6 (~67%) so we only have 1/3 of the capacity available for actual data storage. That matches what Matt shows in the spreadsheet.

With `zpool list` reporting 40 TiB available, there should be 40 TiB * 1/3 = 13.2 TIB of space so there is a delta in space (0.3 TiB) which I can't account for. It isn't a rounding error since using the exact number of bytes (`zpool list` reports 43980465111040 for size and `zfs list` reports 5679360 used and 14146542851840 available) gives a difference of 0.46 TiB.

Also, the slop space would use 3.5% (1.4 TiB) of the available space so I assume that the output from `zfs list` isn't taking that into account? I'm not sure how to factor in the `spa_asize_inflation` into the calculation.

**Comment by ahrens** [comment]
> Does it always 128 KiB or does it really use the recordsize?

The `AVAIL` in `zfs list` is calculated assuming 128K blocks, regardless of the recordsize setting.
